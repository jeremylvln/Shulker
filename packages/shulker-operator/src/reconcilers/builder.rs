use std::fmt::Debug;
use tracing::*;

use kube::{
    api::{DeleteParams, Patch, PatchParams, PostParams},
    Api, ResourceExt,
};
use serde::{de::DeserializeOwned, Serialize};

#[async_trait::async_trait]
pub trait ResourceBuilder {
    type OwnerType: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug;
    type ResourceType: kube::Resource<DynamicType = ()>
        + Clone
        + Serialize
        + DeserializeOwned
        + Debug;

    fn name(owner: &Self::OwnerType) -> String;
    fn api(&self, owner: &Self::OwnerType) -> Api<Self::ResourceType>;

    fn is_needed(&self, _owner: &Self::OwnerType) -> bool {
        true
    }

    fn is_recreation_needed(
        _owner: &Self::OwnerType,
        _existing_resource: &Self::ResourceType,
    ) -> bool {
        false
    }

    async fn build(
        &self,
        owner: &Self::OwnerType,
        name: &str,
        existing_resource: Option<&Self::ResourceType>,
    ) -> Result<Self::ResourceType, anyhow::Error>;
}

async fn get_existing<
    R: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug,
>(
    api: &Api<R>,
    name: &str,
) -> Result<Option<R>, anyhow::Error> {
    api.get(name).await.map(|r| Some(r)).or_else(|e| match e {
        kube::Error::Api(e) => match e.code {
            404 => Ok(None),
            _ => Err(e.into()),
        },
        e => Err(e.into()),
    })
}

fn set_controller_reference<
    O: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug,
    R: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug,
>(
    owner: &O,
    resource: &mut R,
) {
    resource
        .owner_references_mut()
        .push(owner.controller_owner_ref(&()).unwrap());
}

pub async fn reconcile_builder<
    O: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug,
    R: kube::Resource<DynamicType = ()> + Clone + Serialize + DeserializeOwned + Debug,
    RB: ResourceBuilder<OwnerType = O, ResourceType = R>,
>(
    builder: &RB,
    owner: &O,
) -> super::Result<Option<R>> {
    let api = builder.api(owner);
    let name = RB::name(owner);

    debug!(
        builder = std::any::type_name::<RB>(),
        owner = std::any::type_name::<O>(),
        owner_name = owner.name_any(),
        target = std::any::type_name::<R>(),
        target_name = name,
        "reconciling builder",
    );

    let mut existing_resource = get_existing(&api, &name)
        .await
        .map_err(|e| super::ReconcilerError::BuilderError(std::any::type_name::<RB>(), e))?;

    if existing_resource.is_some() {
        debug!(
            builder = std::any::type_name::<RB>(),
            owner = std::any::type_name::<O>(),
            owner_name = owner.name_any(),
            target = std::any::type_name::<R>(),
            target_name = name,
            "found existing resource",
        );

        let delete_existing = || async {
            api.delete(&name, &DeleteParams::default())
                .await
                .map_err(|e| {
                    super::ReconcilerError::BuilderError(std::any::type_name::<RB>(), e.into())
                })
        };

        if !builder.is_needed(owner) {
            debug!(
                builder = std::any::type_name::<RB>(),
                owner = std::any::type_name::<O>(),
                owner_name = owner.name_any(),
                target = std::any::type_name::<R>(),
                target_name = name,
                "existing resource is not needed anymore, deleting",
            );

            delete_existing().await?;
            return Ok(None);
        } else if RB::is_recreation_needed(owner, existing_resource.as_ref().unwrap()) {
            debug!(
                builder = std::any::type_name::<RB>(),
                owner = std::any::type_name::<O>(),
                owner_name = owner.name_any(),
                target = std::any::type_name::<R>(),
                target_name = name,
                "existing resource needs to be recreated, deleting",
            );

            delete_existing().await?;
            existing_resource = None;
        }
    }

    if !builder.is_needed(owner) {
        return Ok(None);
    }

    let mut new_resource = builder
        .build(owner, &name, existing_resource.as_ref())
        .await
        .map_err(|e| super::ReconcilerError::BuilderError(std::any::type_name::<RB>(), e))?;

    let updated_resource = if let Some(existing_resource) = existing_resource {
        debug!(
            builder = std::any::type_name::<RB>(),
            owner = std::any::type_name::<O>(),
            owner_name = owner.name_any(),
            target = std::any::type_name::<R>(),
            target_name = name,
            "patching existing resource",
        );

        for (key, value) in existing_resource.annotations().iter() {
            if !new_resource.annotations().contains_key(key) {
                new_resource
                    .annotations_mut()
                    .insert(key.clone(), value.clone());
            }
        }

        api.patch(
            &name,
            &PatchParams::apply("shulker-operator").force(),
            &Patch::Apply(&new_resource),
        )
        .await
        .map_err(|e| super::ReconcilerError::BuilderError(std::any::type_name::<RB>(), e.into()))?
    } else {
        debug!(
            builder = std::any::type_name::<RB>(),
            owner = std::any::type_name::<O>(),
            owner_name = owner.name_any(),
            target = std::any::type_name::<R>(),
            target_name = name,
            "creating new resource",
        );

        set_controller_reference(owner, &mut new_resource);
        api.create(
            &PostParams {
                dry_run: false,
                field_manager: Some("shulker-operator".to_string()),
            },
            &new_resource,
        )
        .await
        .map_err(|e| super::ReconcilerError::BuilderError(std::any::type_name::<RB>(), e.into()))?
    };

    Ok(Some(updated_resource))
}
