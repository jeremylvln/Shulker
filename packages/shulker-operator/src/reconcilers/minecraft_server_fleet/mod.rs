use std::{collections::BTreeMap, sync::Arc, time::Duration};

use futures::StreamExt;
use google_agones_crds::v1::game_server::GameServer;
use k8s_openapi::api::core::v1::ConfigMap;
use kube::{
    api::{ListParams, PatchParams},
    runtime::{
        controller::Action,
        finalizer::{finalizer, Event as Finalizer},
        watcher::Config,
        Controller,
    },
    Api, Client, ResourceExt,
};
use tracing::*;

use shulker_crds::{
    condition::{ConditionStatus, HasConditions},
    v1alpha1::minecraft_server_fleet::{MinecraftServerFleet, MinecraftServerFleetStatus},
};

use self::{
    config_map::ConfigMapBuilder, fleet::FleetBuilder, fleet_autoscaler::FleetAutoscalerBuilder,
};

use super::{builder::reconcile_builder, status::patch_status, ReconcilerError, Result};

mod config_map;
mod fleet;
mod fleet_autoscaler;

static FINALIZER: &str = "minecraftserverfleets.shulkermc.io";

struct MinecraftServerFleetReconciler {
    client: kube::Client,

    // Builders
    config_map_builder: ConfigMapBuilder,
    fleet_builder: FleetBuilder,
    fleet_autoscaler_builder: FleetAutoscalerBuilder,
}

impl MinecraftServerFleetReconciler {
    async fn reconcile(
        &self,
        api: Api<MinecraftServerFleet>,
        minecraft_server_fleet: Arc<MinecraftServerFleet>,
    ) -> Result<Action> {
        reconcile_builder(&self.config_map_builder, minecraft_server_fleet.as_ref()).await?;
        let fleet = reconcile_builder(&self.fleet_builder, minecraft_server_fleet.as_ref()).await?;
        reconcile_builder(
            &self.fleet_autoscaler_builder,
            minecraft_server_fleet.as_ref(),
        )
        .await?;

        if let Some(fleet) = &fleet {
            if let Some(fleet_status) = &fleet.status {
                let mut minecraft_server_fleet = minecraft_server_fleet.as_ref().clone();
                if minecraft_server_fleet.status.is_none() {
                    minecraft_server_fleet.status = Some(MinecraftServerFleetStatus::default());
                }

                let status = minecraft_server_fleet.status.as_mut().unwrap();

                status.replicas = fleet_status.replicas;
                status.ready_replicas = fleet_status.ready_replicas;
                status.allocated_replicas = fleet_status.allocated_replicas;

                if status.ready_replicas > 0 || status.allocated_replicas > 0 {
                    status.set_condition(
                        "Available".to_string(),
                        ConditionStatus::True,
                        "AtLeastOneReadyOrAllocated".to_string(),
                        "One or more servers are ready or allocated".to_string(),
                    );
                } else {
                    status.set_condition(
                        "Available".to_string(),
                        ConditionStatus::False,
                        "NoneReady".to_string(),
                        "No server are ready".to_string(),
                    );
                };

                patch_status(
                    &api,
                    &PatchParams::apply("shulker-operator"),
                    &minecraft_server_fleet,
                )
                .await?;
            }
        }

        Ok(Action::requeue(Duration::from_secs(5 * 60)))
    }

    async fn cleanup(&self, minecraft_server_fleet: Arc<MinecraftServerFleet>) -> Result<Action> {
        info!(
            name = minecraft_server_fleet.name_any(),
            namespace = minecraft_server_fleet.namespace(),
            "cleaning up MinecraftServerFleet",
        );

        Ok(Action::await_change())
    }

    fn get_common_labels(
        minecraft_server_fleet: &MinecraftServerFleet,
    ) -> BTreeMap<String, String> {
        BTreeMap::from([
            (
                "app.kubernetes.io/name".to_string(),
                minecraft_server_fleet.name_any(),
            ),
            (
                "app.kubernetes.io/component".to_string(),
                "minecraftserver".to_string(),
            ),
            (
                "minecraftcluster.shulkermc.io/name".to_string(),
                minecraft_server_fleet.spec.cluster_ref.name.clone(),
            ),
            (
                "minecraftserverfleet.shulkermc.io/name".to_string(),
                minecraft_server_fleet.name_any(),
            ),
        ])
    }
}

#[instrument(skip(ctx, minecraft_server_fleet))]
async fn reconcile(
    minecraft_server_fleet: Arc<MinecraftServerFleet>,
    ctx: Arc<MinecraftServerFleetReconciler>,
) -> Result<Action> {
    let ns = minecraft_server_fleet.namespace().unwrap();
    let minecraft_server_fleets_api: Api<MinecraftServerFleet> =
        Api::namespaced(ctx.client.clone(), &ns);

    info!(
        name = minecraft_server_fleet.name_any(),
        namespace = ns,
        "reconciling MinecraftServerFleet",
    );

    finalizer(
        &minecraft_server_fleets_api,
        FINALIZER,
        minecraft_server_fleet,
        |event| async {
            match event {
                Finalizer::Apply(minecraft_server_fleet) => {
                    ctx.reconcile(
                        minecraft_server_fleets_api.clone(),
                        minecraft_server_fleet.clone(),
                    )
                    .await
                }
                Finalizer::Cleanup(minecraft_server_fleet) => {
                    ctx.cleanup(minecraft_server_fleet.clone()).await
                }
            }
        },
    )
    .await
    .map_err(|e| ReconcilerError::FinalizerError(Box::new(e)))
}

fn error_policy(
    _minecraft_server_fleet: Arc<MinecraftServerFleet>,
    error: &ReconcilerError,
    _ctx: Arc<MinecraftServerFleetReconciler>,
) -> Action {
    warn!("reconcile failed: {:?}", error);
    // ctx.metrics.reconcile_failure(&minecraft_server_fleet, error);
    Action::requeue(Duration::from_secs(5))
}

pub async fn run(client: Client) {
    let minecraft_server_fleets_api = Api::<MinecraftServerFleet>::all(client.clone());
    if let Err(e) = minecraft_server_fleets_api
        .list(&ListParams::default().limit(1))
        .await
    {
        error!("CRD is not queryable; {e:?}. Is the CRD installed?");
        std::process::exit(1);
    }

    let context = MinecraftServerFleetReconciler {
        client: client.clone(),
        config_map_builder: ConfigMapBuilder::new(client.clone()),
        fleet_builder: FleetBuilder::new(client.clone()),
        fleet_autoscaler_builder: FleetAutoscalerBuilder::new(client.clone()),
    };

    Controller::new(
        minecraft_server_fleets_api,
        Config::default().any_semantic(),
    )
    .owns(
        Api::<ConfigMap>::all(client.clone()),
        Config::default().any_semantic(),
    )
    .owns(
        Api::<GameServer>::all(client.clone()),
        Config::default().any_semantic(),
    )
    .shutdown_on_signal()
    .run(reconcile, error_policy, context.into())
    .filter_map(|x| async move { std::result::Result::ok(x) })
    .for_each(|_| futures::future::ready(()))
    .await;
}
