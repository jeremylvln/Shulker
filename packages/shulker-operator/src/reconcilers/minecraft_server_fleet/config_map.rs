use k8s_openapi::api::core::v1::ConfigMap;
use kube::core::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use crate::reconcilers::builder::ResourceBuilder;
use shulker_crds::v1alpha1::minecraft_server_fleet::MinecraftServerFleet;

use super::MinecraftServerFleetReconciler;

pub struct ConfigMapBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl ResourceBuilder for ConfigMapBuilder {
    type OwnerType = MinecraftServerFleet;
    type ResourceType = ConfigMap;

    fn name(minecraft_server_fleet: &Self::OwnerType) -> String {
        format!("{}-config", minecraft_server_fleet.name_any())
    }

    fn is_updatable() -> bool {
        true
    }

    fn api(&self, minecraft_server_fleet: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(
            self.client.clone(),
            minecraft_server_fleet.namespace().as_ref().unwrap(),
        )
    }

    fn is_needed(&self, _minecraft_server_fleet: &Self::OwnerType) -> bool {
        true
    }

    async fn create(
        &self,
        minecraft_server_fleet: &Self::OwnerType,
        name: &str,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let config_map = ConfigMap {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(minecraft_server_fleet.namespace().unwrap().clone()),
                labels: Some(
                    MinecraftServerFleetReconciler::get_common_labels(minecraft_server_fleet)
                        .into_iter()
                        .collect(),
                ),
                ..ObjectMeta::default()
            },
            ..ConfigMap::default()
        };

        Ok(config_map)
    }

    async fn update(
        &self,
        minecraft_server_fleet: &Self::OwnerType,
        config_map: &mut Self::ResourceType,
    ) -> Result<(), anyhow::Error> {
        config_map.data = Some(
            crate::reconcilers::minecraft_server::config_map::ConfigMapBuilder::get_data_from_spec(
                &minecraft_server_fleet.spec.template.spec.config,
            ),
        );
        Ok(())
    }
}

impl ConfigMapBuilder {
    pub fn new(client: Client) -> Self {
        ConfigMapBuilder { client }
    }
}

#[cfg(test)]
mod tests {
    use crate::reconcilers::{
        builder::ResourceBuilder,
        minecraft_server_fleet::fixtures::{create_client_mock, TEST_SERVER_FLEET},
    };

    #[test]
    fn name_contains_fleet_name() {
        // W
        let name = super::ConfigMapBuilder::name(&TEST_SERVER_FLEET);

        // T
        assert_eq!(name, "my-server-config");
    }

    #[tokio::test]
    async fn create_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::ConfigMapBuilder::new(client);

        // W
        let config_map = builder
            .create(&TEST_SERVER_FLEET, "my-server-config")
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(config_map);
    }

    #[tokio::test]
    async fn update_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::ConfigMapBuilder::new(client);
        let mut config_map = builder
            .create(&TEST_SERVER_FLEET, "my-server-config")
            .await
            .unwrap();

        // W
        builder
            .update(&TEST_SERVER_FLEET, &mut config_map)
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(config_map);
    }
}
