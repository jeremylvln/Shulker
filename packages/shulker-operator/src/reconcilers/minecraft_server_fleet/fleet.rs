use std::collections::BTreeMap;

use google_agones_crds::v1::fleet::FleetTemplate;
use k8s_openapi::api::apps::v1::DeploymentStrategy;
use k8s_openapi::apimachinery::pkg::apis::meta::v1::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use crate::reconcilers::builder::ResourceBuilder;
use crate::resources::resourceref_resolver::ResourceRefResolver;
use google_agones_crds::v1::fleet::Fleet;
use google_agones_crds::v1::fleet::FleetSpec;
use shulker_crds::v1alpha1::minecraft_server::MinecraftServer;
use shulker_crds::v1alpha1::minecraft_server::MinecraftServerConfigurationSpec;
use shulker_crds::v1alpha1::minecraft_server::MinecraftServerSpec;
use shulker_crds::v1alpha1::minecraft_server_fleet::MinecraftServerFleet;

use super::config_map::ConfigMapBuilder;
use super::MinecraftServerFleetReconciler;

pub struct FleetBuilder {
    client: Client,
    resourceref_resolver: ResourceRefResolver,
}

#[async_trait::async_trait]
impl ResourceBuilder for FleetBuilder {
    type OwnerType = MinecraftServerFleet;
    type ResourceType = Fleet;

    fn name(minecraft_server_fleet: &Self::OwnerType) -> String {
        minecraft_server_fleet.name_any()
    }

    fn is_updatable() -> bool {
        true
    }

    fn api(&self, minecraft_server_fleet: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(
            self.client.clone(),
            minecraft_server_fleet.namespace().as_ref().unwrap(),
        )
    }

    fn is_needed(&self, _minecraft_server_fleet: &Self::OwnerType) -> bool {
        true
    }

    async fn create(
        &self,
        minecraft_server_fleet: &Self::OwnerType,
        name: &str,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let fleet = Fleet {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(minecraft_server_fleet.namespace().unwrap().clone()),
                labels: Some(
                    MinecraftServerFleetReconciler::get_common_labels(minecraft_server_fleet)
                        .into_iter()
                        .collect(),
                ),
                ..ObjectMeta::default()
            },
            spec: FleetSpec::default(),
            status: None,
        };

        Ok(fleet)
    }

    async fn update(
        &self,
        minecraft_server_fleet: &Self::OwnerType,
        fleet: &mut Self::ResourceType,
    ) -> Result<(), anyhow::Error> {
        let fake_mincraft_server = MinecraftServer {
            metadata: ObjectMeta {
                namespace: minecraft_server_fleet.namespace(),
                name: Some(minecraft_server_fleet.name_any()),
                ..ObjectMeta::default()
            },
            spec: MinecraftServerSpec {
                cluster_ref: minecraft_server_fleet
                    .spec
                    .template
                    .spec
                    .cluster_ref
                    .clone(),
                config: MinecraftServerConfigurationSpec {
                    existing_config_map_name: Some(ConfigMapBuilder::name(minecraft_server_fleet)),
                    ..MinecraftServerConfigurationSpec::default()
                },
                ..minecraft_server_fleet.spec.template.spec.clone()
            },
            status: None,
        };

        let game_server_spec = crate::reconcilers::minecraft_server::gameserver::GameServerBuilder::get_game_server_spec(
            &self.resourceref_resolver,
            &fake_mincraft_server,
        ).await?;
        let replicas = match &minecraft_server_fleet.spec.autoscaling {
            Some(_) => 0,
            None => minecraft_server_fleet.spec.replicas as i32,
        };

        let mut labels = MinecraftServerFleetReconciler::get_common_labels(minecraft_server_fleet);
        labels.extend(
            game_server_spec
                .template
                .metadata
                .as_ref()
                .map_or_else(BTreeMap::new, |metadata| {
                    metadata.labels.clone().unwrap_or_default()
                }),
        );

        let mut annotations = BTreeMap::<String, String>::new();
        annotations.extend(
            game_server_spec
                .template
                .metadata
                .as_ref()
                .map_or_else(BTreeMap::new, |metadata| {
                    metadata.annotations.clone().unwrap_or_default()
                }),
        );

        if let Some(metadata) = &minecraft_server_fleet.spec.template.metadata {
            if let Some(additional_labels) = metadata.labels.clone() {
                labels.extend(additional_labels);
            }

            if let Some(additional_annotations) = metadata.annotations.clone() {
                annotations.extend(additional_annotations);
            }
        }

        fleet.spec = FleetSpec {
            replicas: Some(replicas),
            strategy: Some(DeploymentStrategy {
                type_: Some("Recreate".to_string()),
                ..DeploymentStrategy::default()
            }),
            scheduling: Some("Packed".to_string()),
            template: FleetTemplate {
                metadata: Some(ObjectMeta {
                    labels: Some(labels),
                    annotations: Some(annotations),
                    ..ObjectMeta::default()
                }),
                spec: game_server_spec,
            },
        };

        Ok(())
    }
}

impl FleetBuilder {
    pub fn new(client: Client) -> Self {
        FleetBuilder {
            client: client.clone(),
            resourceref_resolver: ResourceRefResolver::new(client.clone()),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::reconcilers::{
        builder::ResourceBuilder,
        minecraft_server_fleet::fixtures::{create_client_mock, TEST_SERVER_FLEET},
    };

    #[test]
    fn name_contains_fleet_name() {
        // W
        let name = super::FleetBuilder::name(&TEST_SERVER_FLEET);

        // T
        assert_eq!(name, "my-server");
    }

    #[tokio::test]
    async fn create_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::FleetBuilder::new(client);

        // W
        let fleet = builder
            .create(&TEST_SERVER_FLEET, "my-server")
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(fleet);
    }

    #[tokio::test]
    async fn update_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::FleetBuilder::new(client);
        let mut fleet = builder
            .create(&TEST_SERVER_FLEET, "my-server")
            .await
            .unwrap();

        // W
        builder
            .update(&TEST_SERVER_FLEET, &mut fleet)
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(fleet);
    }

    #[tokio::test]
    async fn update_should_merge_labels() {
        // G
        let client = create_client_mock();
        let builder = super::FleetBuilder::new(client);
        let mut fleet = builder
            .create(&TEST_SERVER_FLEET, "my-server")
            .await
            .unwrap();

        // W
        builder
            .update(&TEST_SERVER_FLEET, &mut fleet)
            .await
            .unwrap();

        // T
        let additional_labels = TEST_SERVER_FLEET
            .spec
            .template
            .metadata
            .as_ref()
            .unwrap()
            .labels
            .as_ref()
            .unwrap();
        additional_labels.iter().for_each(|(key, value)| {
            assert_eq!(
                fleet
                    .spec
                    .template
                    .metadata
                    .as_ref()
                    .unwrap()
                    .labels
                    .as_ref()
                    .unwrap()
                    .get(key)
                    .unwrap(),
                value
            );
        });
    }

    #[tokio::test]
    async fn update_should_merge_annotations() {
        // G
        let client = create_client_mock();
        let builder = super::FleetBuilder::new(client);
        let mut fleet = builder
            .create(&TEST_SERVER_FLEET, "my-server")
            .await
            .unwrap();

        // W
        builder
            .update(&TEST_SERVER_FLEET, &mut fleet)
            .await
            .unwrap();

        // T
        let additional_annotations = TEST_SERVER_FLEET
            .spec
            .template
            .metadata
            .as_ref()
            .unwrap()
            .annotations
            .as_ref()
            .unwrap();
        additional_annotations.iter().for_each(|(key, value)| {
            assert_eq!(
                fleet
                    .spec
                    .template
                    .metadata
                    .as_ref()
                    .unwrap()
                    .annotations
                    .as_ref()
                    .unwrap()
                    .get(key)
                    .unwrap(),
                value
            );
        });
    }

    #[tokio::test]
    async fn update_should_use_recreate_strategy() {
        // G
        let client = create_client_mock();
        let builder = super::FleetBuilder::new(client);
        let mut fleet = builder
            .create(&TEST_SERVER_FLEET, "my-server")
            .await
            .unwrap();

        // W
        builder
            .update(&TEST_SERVER_FLEET, &mut fleet)
            .await
            .unwrap();

        // T
        assert_eq!(
            fleet
                .spec
                .strategy
                .as_ref()
                .unwrap()
                .type_
                .as_ref()
                .unwrap(),
            "Recreate"
        );
    }
}
