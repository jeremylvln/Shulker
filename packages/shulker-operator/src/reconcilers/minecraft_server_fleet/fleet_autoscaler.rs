use k8s_openapi::apimachinery::pkg::apis::meta::v1::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use crate::reconcilers::builder::ResourceBuilder;
use google_agones_crds::v1::fleet_autoscaler::FleetAutoscaler;
use google_agones_crds::v1::fleet_autoscaler::FleetAutoscalerSpec;
use shulker_crds::v1alpha1::minecraft_server_fleet::MinecraftServerFleet;

use super::fleet::FleetBuilder;
use super::MinecraftServerFleetReconciler;

pub struct FleetAutoscalerBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl ResourceBuilder for FleetAutoscalerBuilder {
    type OwnerType = MinecraftServerFleet;
    type ResourceType = FleetAutoscaler;

    fn name(minecraft_server_fleet: &Self::OwnerType) -> String {
        minecraft_server_fleet.name_any()
    }

    fn is_updatable() -> bool {
        true
    }

    fn api(&self, minecraft_server_fleet: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(
            self.client.clone(),
            minecraft_server_fleet.namespace().as_ref().unwrap(),
        )
    }

    fn is_needed(&self, minecraft_server_fleet: &Self::OwnerType) -> bool {
        match &minecraft_server_fleet.spec.autoscaling {
            Some(autoscaling) => autoscaling.agones_policy.is_some(),
            None => false,
        }
    }

    async fn create(
        &self,
        minecraft_server_fleet: &Self::OwnerType,
        name: &str,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let fleet_autoscaler = FleetAutoscaler {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(minecraft_server_fleet.namespace().unwrap().clone()),
                labels: Some(
                    MinecraftServerFleetReconciler::get_common_labels(minecraft_server_fleet)
                        .into_iter()
                        .collect(),
                ),
                ..ObjectMeta::default()
            },
            spec: FleetAutoscalerSpec::default(),
            status: None,
        };

        Ok(fleet_autoscaler)
    }

    async fn update(
        &self,
        minecraft_server_fleet: &Self::OwnerType,
        fleet_autoscaler: &mut Self::ResourceType,
    ) -> Result<(), anyhow::Error> {
        fleet_autoscaler.spec = FleetAutoscalerSpec {
            fleet_name: FleetBuilder::name(minecraft_server_fleet),
            policy: minecraft_server_fleet
                .spec
                .autoscaling
                .as_ref()
                .unwrap()
                .agones_policy
                .as_ref()
                .unwrap()
                .clone(),
        };
        Ok(())
    }
}

impl FleetAutoscalerBuilder {
    pub fn new(client: Client) -> Self {
        FleetAutoscalerBuilder { client }
    }
}
