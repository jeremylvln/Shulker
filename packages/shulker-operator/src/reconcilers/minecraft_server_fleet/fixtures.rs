use std::collections::BTreeMap;

use google_agones_crds::v1::fleet_autoscaler::{
    FleetAutoscalerPolicy, FleetAutoscalerPolicyBufferSpec, FleetAutoscalerPolicySpec,
};
use http::{Request, Response};
use k8s_openapi::{api::core::v1::EnvVar, apimachinery::pkg::util::intstr::IntOrString};
use kube::client::Body;
use kube::{core::ObjectMeta, Client};
use lazy_static::lazy_static;
use shulker_crds::{
    resourceref::ResourceRefSpec,
    schemas::{FleetAutoscalingSpec, TemplateSpec},
    v1alpha1::{
        minecraft_cluster::MinecraftClusterRef,
        minecraft_server::{
            MinecraftServerConfigurationLifecycleStrategy,
            MinecraftServerConfigurationProxyForwardingMode, MinecraftServerConfigurationSpec,
            MinecraftServerPodOverridesSpec, MinecraftServerSpec, MinecraftServerVersion,
            MinecraftServerVersionSpec,
        },
        minecraft_server_fleet::{MinecraftServerFleet, MinecraftServerFleetSpec},
    },
};

lazy_static! {
    pub static ref TEST_SERVER_FLEET: MinecraftServerFleet = MinecraftServerFleet {
        metadata: ObjectMeta {
            namespace: Some("default".to_string()),
            name: Some("my-server".to_string()),
            ..ObjectMeta::default()
        },
        spec: MinecraftServerFleetSpec {
            cluster_ref: MinecraftClusterRef::new("my-cluster".to_string()),
            replicas: 3,
            template: TemplateSpec {
                metadata: Some(ObjectMeta {
                    labels: Some(BTreeMap::from([(
                        "test-label/shulkermc.io".to_string(),
                        "my-value".to_string()
                    )])),
                    annotations: Some(BTreeMap::from([(
                        "test-annotation/shulkermc.io".to_string(),
                        "my-value".to_string()
                    )])),
                    ..ObjectMeta::default()
                }),
                spec: MinecraftServerSpec {
                    cluster_ref: MinecraftClusterRef::new("my-cluster".to_string()),
                    tags: vec!["lobby".to_string()],
                    version: MinecraftServerVersionSpec {
                        channel: MinecraftServerVersion::Paper,
                        name: "1.20.1".to_string(),
                        ..MinecraftServerVersionSpec::default()
                    },
                    config: MinecraftServerConfigurationSpec {
                        existing_config_map_name: None,
                        world: Some(ResourceRefSpec {
                            url: Some("https://example.com/my_world.tar.gz".to_string()),
                            ..ResourceRefSpec::default()
                        }),
                        plugins: Some(vec![ResourceRefSpec {
                            url: Some("https://example.com/my_plugin.jar".to_string()),
                            ..ResourceRefSpec::default()
                        }]),
                        patches: Some(vec![ResourceRefSpec {
                            url: Some("https://example.com/my_patch.tar.gz".to_string()),
                            ..ResourceRefSpec::default()
                        }]),
                        max_players: 42,
                        disable_nether: false,
                        disable_end: true,
                        server_properties: None,
                        proxy_forwarding_mode:
                            MinecraftServerConfigurationProxyForwardingMode::Velocity,
                        lifecycle_strategy:
                            MinecraftServerConfigurationLifecycleStrategy::AllocateWhenNotEmpty,
                        skip_agent_download: false
                    },
                    pod_overrides: Some(MinecraftServerPodOverridesSpec {
                        image: None,
                        env: Some(vec![EnvVar {
                            name: "EXTRA_ENV".to_string(),
                            value: Some("my_value".to_string()),
                            ..EnvVar::default()
                        }]),
                        resources: None,
                        affinity: None,
                        node_selector: Some(BTreeMap::from([(
                            "beta.kubernetes.io/os".to_string(),
                            "linux".to_string()
                        )])),
                        tolerations: None,
                        service_account_name: None,
                        volume_mounts: None,
                        volumes: None,
                        ports: None,
                        health: None,
                    })
                },
            },
            autoscaling: Some(FleetAutoscalingSpec {
                agones_policy: Some(FleetAutoscalerPolicySpec {
                    type_: FleetAutoscalerPolicy::Buffer,
                    buffer: Some(FleetAutoscalerPolicyBufferSpec {
                        max_replicas: 8,
                        min_replicas: 3,
                        buffer_size: IntOrString::Int(2),
                    }),
                    webhook: None,
                })
            }),
        },
        status: None,
    };
}

pub fn create_client_mock() -> Client {
    let (mock_service, _) = tower_test::mock::pair::<Request<Body>, Response<Body>>();
    Client::new(mock_service, "default")
}
