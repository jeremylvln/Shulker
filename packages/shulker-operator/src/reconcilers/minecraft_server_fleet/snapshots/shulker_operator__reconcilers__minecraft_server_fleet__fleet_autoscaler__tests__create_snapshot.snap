---
source: packages/shulker-operator/src/reconcilers/minecraft_server_fleet/fleet_autoscaler.rs
expression: fleet_autoscaler
---
apiVersion: autoscaling.agones.dev/v1
kind: FleetAutoscaler
metadata:
  labels:
    app.kubernetes.io/component: minecraftserver
    app.kubernetes.io/name: my-server
    minecraftcluster.shulkermc.io/name: my-cluster
    minecraftserverfleet.shulkermc.io/name: my-server
  name: my-server
  namespace: default
spec:
  fleetName: ""
  policy:
    type: Buffer

